### ISSUE TITLE
two competing `SELECT FOR UPDATE` statements allowed for the same row after `CREATE TABLE`

### ISSUE THREAD

**Issue by jabielecki** [cockroachdb/cockroach#83571]
**Describe the problem**

When two competing transactions issue `SELECT FOR UPDATE` there is a possibility that the same exact row is selected by both statements.

TL;DR:

```sql
CREATE TABLE tab ( id INTEGER, b BOOL );
INSERT INTO tab(id, b) VALUES (1, true);
BEGIN;
SELECT * FROM tab WHERE id = 1 FOR UPDATE ;
```
```sql
SELECT * FROM tab WHERE id = 1 FOR UPDATE ;  // <-- can succeed in a separate session after about 6~7 seconds from CREATE TABLE...
```

**To Reproduce**

The behavior was quite intermittent in the beginning for me, but I was able to nail down a repro.

```sh
$ docker run --rm --name crdb -d cockroachdb/cockroach:v22.1.2 start-single-node --insecure

$ docker exec -it crdb ./cockroach sql --insecure
```

Here you can paste all this text. For this fragment it is actually crucial that the seemingly redundant `COMMIT` is there, and that the statements exec _immediately_ one after another. If I add one second pauses between commands, it's no longer reproducible.

```sql
CREATE TABLE tab ( id INTEGER, b BOOL );
INSERT INTO tab(id, b) VALUES (1, true);
COMMIT;
BEGIN;
SET LOCAL application_name='a1';
SHOW ALL TRANSACTIONS;
SELECT * FROM tab WHERE id = 1 FOR UPDATE;
SHOW ALL TRANSACTIONS;
```

Press Enter. This produces:

```txt
CREATE TABLE


Time: 9ms total (execution 9ms / network 0ms)

INSERT 1


Time: 23ms total (execution 23ms / network 0ms)

ERROR: there is no transaction in progress
BEGIN


Time: 0ms total (execution 0ms / network 0ms)

SET


Time: 0ms total (execution 0ms / network 0ms)

  node_id |                txn_id                | application_name | num_stmts | num_retries | num_auto_retries
----------+--------------------------------------+------------------+-----------+-------------+-------------------
        1 | 5e5894b4-8161-4361-bc3f-853c62d433d2 | a1               |         4 |           0 |                0
(1 row)


Time: 2ms total (execution 2ms / network 0ms)

  id |  b
-----+-------
   1 | true
(1 row)


Time: 1ms total (execution 1ms / network 0ms)

  node_id |                txn_id                | application_name | num_stmts | num_retries | num_auto_retries
----------+--------------------------------------+------------------+-----------+-------------+-------------------
        1 | 5e5894b4-8161-4361-bc3f-853c62d433d2 | a1               |         8 |           0 |                0
(1 row)


Time: 2ms total (execution 1ms / network 0ms)
```

Keep the session open, go to another terminal.

```
$ docker exec -it crdb ./cockroach sql --insecure
BEGIN;
SET LOCAL application_name='a2';
SHOW ALL TRANSACTIONS;
SELECT * FROM tab WHERE id = 1 FOR UPDATE NOWAIT;
SHOW ALL TRANSACTIONS;
```

This succeeds and produces:

```txt
BEGIN


Time: 0ms total (execution 0ms / network 0ms)

SET


Time: 0ms total (execution 0ms / network 0ms)

  node_id |                txn_id                | application_name | num_stmts | num_retries | num_auto_retries
----------+--------------------------------------+------------------+-----------+-------------+-------------------
        1 | 5e5894b4-8161-4361-bc3f-853c62d433d2 | a1               |        10 |           0 |                0
        1 | c8e20d73-e706-4463-9d6f-208571ab04f2 | a2               |         4 |           0 |                0
(2 rows)


Time: 2ms total (execution 2ms / network 0ms)

  id |  b
-----+-------
   1 | true
(1 row)


Time: 1ms total (execution 1ms / network 0ms)

  node_id |                txn_id                | application_name | num_stmts | num_retries | num_auto_retries
----------+--------------------------------------+------------------+-----------+-------------+-------------------
        1 | 5e5894b4-8161-4361-bc3f-853c62d433d2 | a1               |        10 |           0 |                0
        1 | c8e20d73-e706-4463-9d6f-208571ab04f2 | a2               |         8 |           0 |                0
(2 rows)


Time: 2ms total (execution 2ms / network 0ms)
```

**Expected behavior**

The `SELECT * FROM tab WHERE id = 1 FOR UPDATE NOWAIT;` above should fail with `ERROR: could not obtain lock on row (rowid)=(...) in tab@tab_pkey`

**Additional data / screenshots**

```
$ docker logs crdb
ls: cannot access 'cockroach-data': No such file or directory
certificate dir "certs" is successfully set up
starting node for the initialization process. This could take a couple seconds...
*
* WARNING: ALL SECURITY CONTROLS HAVE BEEN DISABLED!
* 
* This mode is intended for non-production testing only.
* 
* In this mode:
* - Your cluster is open to any client that can access any of your IP addresses.
* - Intruders with access to your machine or network can observe client-server traffic.
* - Intruders can log in without password and read or write any data in the cluster.
* - Intruders can consume all your server's resources and cause unavailability.
*
*
* INFO: To start a secure server without mandating TLS for clients,
* consider --accept-sql-without-tls instead. For other options, see:
* 
* - https://go.crdb.dev/issue-v/53404/v22.1
* - https://www.cockroachlabs.com/docs/v22.1/secure-a-cluster.html
*
init node successfully started
CockroachDB node starting at 2022-06-29 08:45:24.602708685 +0000 UTC (took 0.6s)
build:               CCL v22.1.2 @ 2022/06/22 15:54:12 (go1.17.11)
webui:               http://127.0.0.1:8080
sql:                 postgresql://root@127.0.0.1:26257/defaultdb?sslmode=disable
sql (JDBC):          jdbc:postgresql://127.0.0.1:26257/defaultdb?sslmode=disable&user=root
RPC client flags:    /cockroach/cockroach <client cmd> --host=127.0.0.1:26257 --insecure
logs:                /cockroach/cockroach-data/logs
temp dir:            /cockroach/cockroach-data/cockroach-temp3690996061
external I/O path:   /cockroach/cockroach-data/extern
store[0]:            path=/cockroach/cockroach-data
storage engine:      pebble
clusterID:           3b299f74-8594-475f-b64b-e2086d1b52c4
status:              initialized new cluster
nodeID:              1
database "defaultdb" already exists
start running init files from /docker-entrypoint-initdb.d
/cockroach/cockroach.sh: ignoring /docker-entrypoint-initdb.d/*

end running init files from /docker-entrypoint-initdb.d
"${start_node_query[@]}"
```

```
$ docker exec -it crdb ./cockroach sql --insecure
#
# Welcome to the CockroachDB SQL shell.
# All statements must be terminated by a semicolon.
# To exit, type: \q.
#
# Server version: CockroachDB CCL v22.1.2 (x86_64-pc-linux-gnu, built 2022/06/22 15:54:12, go1.17.11) (same version as client)
# Cluster ID: 3b299f74-8594-475f-b64b-e2086d1b52c4
#
# Enter \? for a brief introduction.
#
```

```
$ docker image inspect cockroachdb/cockroach:v22.1.2 | head
[
    {
        "Id": "sha256:2ed6db8a067e36542a632a31f0ba76c3eee5e616605369418bc05a0fceab42ae",
        "RepoTags": [
            "cockroachdb/cockroach:latest",
            "cockroachdb/cockroach:v22.1.2"
        ],
        "RepoDigests": [
            "cockroachdb/cockroach@sha256:04a55dfae8e10e9b2c517fd6fd7aa76574895f76b4e2098f287b542de6db3766"
        ],
```

**Environment:**
 - CockroachDB CCL v22.1.2 (x86_64-pc-linux-gnu, built 2022/06/22 15:54:12, go1.17.11)
 - Server OS: Ubuntu 22.04 LTS
 - Client `cockroach sql` from the same docker image as the server

**Additional context**

Two transactions proceed on the exact same row in parallel, resulting in breaking business logic down the line.

Can be also reproduced without a `NOWAIT` flag on `SELECT FOR UPDATE`.

Can be also reproduced when the `id` column is a `PRIMARY KEY`.

Jira issue: CRDB-17146

**Comment by michae2** [comment]
Thanks for the bug report, @jabielecki. Unfortunately I haven't been able to reproduce it yet.

So far, using:
- CockroachDB CCL v22.1.2 (x86_64-apple-darwin21.5.0, built 2022/06/30 01:07:57, go1.17.11)
- Server OS: MacOS 12.4
- `cockroach demo` and `cockroach sql`

I get a `could not obtain lock on row` error from the second transaction:

```
demo@127.0.0.1:26257/defaultdb> BEGIN;
SET LOCAL application_name='a2';
SHOW ALL TRANSACTIONS;
SELECT * FROM tab WHERE id = 1 FOR UPDATE NOWAIT;
SHOW ALL TRANSACTIONS;
BEGIN


Time: 0ms total (execution 0ms / network 0ms)

SET


Time: 0ms total (execution 0ms / network 0ms)

  node_id |                txn_id                | application_name | num_stmts | num_retries | num_auto_retries
----------+--------------------------------------+------------------+-----------+-------------+-------------------
        1 | 17932bea-2ac5-47c7-9750-f6dad39cec47 | a1               |         9 |           0 |                0
        1 | e1dd3b2e-396a-457f-9b6b-e3766c5af07a | a2               |         4 |           0 |                0
(2 rows)


Time: 2ms total (execution 2ms / network 0ms)

ERROR: could not obtain lock on row (rowid)=(774902312778072065) in tab@tab_pkey
SQLSTATE: 55P03
ERROR: current transaction is aborted, commands ignored until end of transaction block
SQLSTATE: 25P02
```

I'll try again tomorrow with Ubuntu, docker, and `cockroach start-single-node`.

**Comment by michae2** [comment]
It's possible this is a timing thing, and I'm just not hitting the right timing. 🤔

**Comment by jabielecki** [comment]
> It's possible this is a timing thing, and I'm just not hitting the right timing.

Yes, definitely timing-dependent. I saw it in a very different setting at first in my app, and it took a lot of fiddling to reproduce it (even if only locally on my Ubuntu box).

**Comment by jabielecki** [comment]
I've learned `cockroach sql` a bit since yesterday and it turns out that the timing is achievable quite easily.

Just paste this:
```
CREATE TABLE tab ( id INTEGER, b BOOL );
\! sleep 0.234
INSERT INTO tab(id, b) VALUES (1, true);
\! sleep 0.234
COMMIT;
\! sleep 0.234
BEGIN;
\! sleep 0.234
SELECT current_timestamp;
SELECT * FROM tab WHERE id = 1 FOR UPDATE;
show all cluster transactions;
\! sleep 0.5 ; echo -e 'select current_timestamp;\n SELECT * FROM tab WHERE id = 1 FOR UPDATE NOWAIT;\n' | cockroach sql --insecure
\! sleep 0.5 ; echo -e 'select current_timestamp;\n SELECT * FROM tab WHERE id = 1 FOR UPDATE NOWAIT;\n' | cockroach sql --insecure
\! sleep 0.5 ; echo -e 'select current_timestamp;\n SELECT * FROM tab WHERE id = 1 FOR UPDATE NOWAIT;\n' | cockroach sql --insecure
\! sleep 0.5 ; echo -e 'select current_timestamp;\n SELECT * FROM tab WHERE id = 1 FOR UPDATE NOWAIT;\n' | cockroach sql --insecure
\! sleep 0.5 ; echo -e 'select current_timestamp;\n SELECT * FROM tab WHERE id = 1 FOR UPDATE NOWAIT;\n' | cockroach sql --insecure
\! sleep 0.5 ; echo -e 'select current_timestamp;\n SELECT * FROM tab WHERE id = 1 FOR UPDATE NOWAIT;\n' | cockroach sql --insecure
\! sleep 0.5 ; echo -e 'select current_timestamp;\n SELECT * FROM tab WHERE id = 1 FOR UPDATE NOWAIT;\n' | cockroach sql --insecure
\! sleep 0.5 ; echo -e 'select current_timestamp;\n SELECT * FROM tab WHERE id = 1 FOR UPDATE NOWAIT;\n' | cockroach sql --insecure
\! sleep 0.5 ; echo -e 'select current_timestamp;\n SELECT * FROM tab WHERE id = 1 FOR UPDATE NOWAIT;\n' | cockroach sql --insecure
\! sleep 0.5 ; echo -e 'select current_timestamp;\n SELECT * FROM tab WHERE id = 1 FOR UPDATE NOWAIT;\n' | cockroach sql --insecure
\! sleep 0.5 ; echo -e 'select current_timestamp;\n SELECT * FROM tab WHERE id = 1 FOR UPDATE NOWAIT;\n' | cockroach sql --insecure
\! sleep 0.5 ; echo -e 'select current_timestamp;\n SELECT * FROM tab WHERE id = 1 FOR UPDATE NOWAIT;\n' | cockroach sql --insecure
show all cluster transactions;
```

Reproducible output:

```txt
CREATE TABLE


Time: 12ms total (execution 12ms / network 0ms)

INSERT 1


Time: 11ms total (execution 11ms / network 0ms)

ERROR: there is no transaction in progress
BEGIN


Time: 0ms total (execution 0ms / network 0ms)

        current_timestamp
---------------------------------
  2022-06-30 13:01:38.551885+00
(1 row)


Time: 0ms total (execution 0ms / network 0ms)

  id |  b
-----+-------
   1 | true
(1 row)


Time: 1ms total (execution 1ms / network 0ms)

  node_id |                txn_id                | application_name | num_stmts | num_retries | num_auto_retries
----------+--------------------------------------+------------------+-----------+-------------+-------------------
        1 | 53ea0d88-f8e1-4c65-a281-2ce3816b234b | $ cockroach sql  |         7 |           0 |                0
(1 row)


Time: 3ms total (execution 2ms / network 0ms)

ERROR: could not obtain lock on row (rowid)=(775040853808676865) in tab@tab_pkey
SQLSTATE: 55P03
ERROR: could not obtain lock on row (rowid)=(775040853808676865) in tab@tab_pkey
SQLSTATE: 55P03
Failed running "sql"
command failed: error in external command: exit status 1
ERROR: could not obtain lock on row (rowid)=(775040853808676865) in tab@tab_pkey
SQLSTATE: 55P03
ERROR: could not obtain lock on row (rowid)=(775040853808676865) in tab@tab_pkey
SQLSTATE: 55P03
Failed running "sql"
command failed: error in external command: exit status 1
ERROR: could not obtain lock on row (rowid)=(775040853808676865) in tab@tab_pkey
SQLSTATE: 55P03
ERROR: could not obtain lock on row (rowid)=(775040853808676865) in tab@tab_pkey
SQLSTATE: 55P03
Failed running "sql"
command failed: error in external command: exit status 1
ERROR: could not obtain lock on row (rowid)=(775040853808676865) in tab@tab_pkey
SQLSTATE: 55P03
ERROR: could not obtain lock on row (rowid)=(775040853808676865) in tab@tab_pkey
SQLSTATE: 55P03
Failed running "sql"
command failed: error in external command: exit status 1
ERROR: could not obtain lock on row (rowid)=(775040853808676865) in tab@tab_pkey
SQLSTATE: 55P03
ERROR: could not obtain lock on row (rowid)=(775040853808676865) in tab@tab_pkey
SQLSTATE: 55P03
Failed running "sql"
command failed: error in external command: exit status 1
ERROR: could not obtain lock on row (rowid)=(775040853808676865) in tab@tab_pkey
SQLSTATE: 55P03
ERROR: could not obtain lock on row (rowid)=(775040853808676865) in tab@tab_pkey
SQLSTATE: 55P03
Failed running "sql"
command failed: error in external command: exit status 1
ERROR: could not obtain lock on row (rowid)=(775040853808676865) in tab@tab_pkey
SQLSTATE: 55P03
ERROR: could not obtain lock on row (rowid)=(775040853808676865) in tab@tab_pkey
SQLSTATE: 55P03
Failed running "sql"
command failed: error in external command: exit status 1
current_timestamp
2022-06-30 13:01:44.520443+00
id	b
1	true
current_timestamp
2022-06-30 13:01:45.243614+00
id	b
1	true
current_timestamp
2022-06-30 13:01:45.94089+00
id	b
1	true
current_timestamp
2022-06-30 13:01:46.667185+00
id	b
1	true
current_timestamp
2022-06-30 13:01:47.3916+00
id	b
1	true
  node_id |                txn_id                | application_name | num_stmts | num_retries | num_auto_retries
----------+--------------------------------------+------------------+-----------+-------------+-------------------
        1 | 53ea0d88-f8e1-4c65-a281-2ce3816b234b | $ cockroach sql  |        21 |           0 |                0
(1 row)

```
For me, the bug happens after approximately 5~6 seconds as in the output above.

Let me know **if you approve of replacing the big contrived (and unreadable) example in the current issue body** with this comment's example.

**Comment by michae2** [comment]
Thanks for the repro, @jabielecki. I am able to reproduce it now. Looks like this happens when the first transaction is open for > 5s before opening the second transaction.

**Comment by michae2** [comment]
The trace of the second transaction includes this log message: https://github.com/cockroachdb/cockroach/blob/b630c08339a65976be8a16eef4965b45c5e338da/pkg/kv/kvserver/concurrency/lock_table_waiter.go#L499-L506

So I think the question is why the second transaction thinks the first is abandoned. And how does the 5s come into play?

**Comment by michae2** [comment]
Notes from investigating with @andreimatei and @nvanbenschoten:
- there's a range split and lease transfer shortly after the `CREATE TABLE`
- we currently don't preserve `FOR UPDATE` locks across lease transfers
- adding a long enough sleep after the `CREATE TABLE` makes this stop reproducing, because the first transaction no longer crosses a lease transfer

**Comment by AlexTalks** [comment]
I think you've covered this above, but `FOR UPDATE` acquires unreplicated locks, so by design if there's a lease transfer, the locks will be dropped.  Having the option to either restart the transaction or acquire a more durable lock may be worth discussion, however.

**Comment by nvanbenschoten** [comment]
I'm going to close this as a duplicate of https://github.com/cockroachdb/cockroach/issues/75456, as that's the work that will be needed to fix this. Thanks for filing @jabielecki!
